<a href="https://colab.research.google.com/github/riinndescartes/Machine-Learning/blob/main/Catherine_P_SVM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import classification_report, confusion_matrix


In [3]:
# Muat data
data = pd.read_csv('bank-additional-full.csv', sep=';')

# Tampilkan 5 data pertama
print(data.head())


   age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

In [5]:
# Pisahkan data fitur dan target
X = data.drop('y', axis=1)
y = data['y']

# Ubah target menjadi 0 dan 1 (yes/no)
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

# Ubah fitur kategorikal menjadi numerik
X = pd.get_dummies(X, drop_first=True)


In [6]:
# Bagi data menjadi data latih dan uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalisasi fitur
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [7]:
# Buat dan latih model SVM
svm_model = SVC(kernel='linear')
svm_model.fit(X_train, y_train)


SVC(kernel='linear')

In [8]:
# Prediksi data uji
y_pred = svm_model.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Confusion Matrix:
[[10725   243]
 [  950   439]]

Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.98      0.95     10968
           1       0.64      0.32      0.42      1389

    accuracy                           0.90     12357
   macro avg       0.78      0.65      0.69     12357
weighted avg       0.89      0.90      0.89     12357



#Latihan



1.   Buat interpretasi dari hasil evaluasi di atas
2.   Latih model SVM menggunakan kernel polyomial, RBF, dan sigmoid. masing-masing, dan tampilkan hasil classification_report untuk setiap kernel. \
  2.1 Berdasarkan f1-score dari hasil evaluasi, tentukan kernel mana yang memberikan hasil terbaik.
3. Untuk kernel poly, coba berbagai nilai degree (misalnya 2, 3, dan 4). Temukan nilai degree yang memberikan hasil terbaik.
4. Untuk kernel rbf, coba beberapa nilai C (misalnya 0.1, 1, dan 10). Tentukan nilai C terbaik berdasarkan f1-score.
5. Dengan menggunakan kernel dan parameter terbaik dari hasil latihan di atas (soal 3 dan 4), tuliskan interpretasi singkat mengenai precision, recall, f1-score, dan confusion matrix yang dihasilkan.
6. Berdasarkan soal 3 dan 4 jelaskan mengapa kernel tertentu mungkin lebih cocok dibandingkan yang lain dalam dataset ini.

https://scikit-learn.org/1.5/modules/svm.html


#1. Interpretasi



**1.1. Confusion Matrix:**

Kita mengukur kinerja model SVM yang menghasilkan *binary classification* menggunakan *confusion matrix*. Matriks ini memetakan kinerja model saat memproses data prediksi dibandingkan kenyataannya.

  [[10725    243]
                      
   [  950      439]]

Merupakan nilai dari Confusion Matrix berikut,

[[TN    FP]
                      
 [  FN     TP]]

**Keterangan Matriks:**
- TP = TRUE POSITIVE, model sukses memprediksi positif ("ya"), karena kenyataan memang positif ("ya").
- TN = TRUE NEGAITVE, model sukses memprediksi negative ("tidak"), karena kenyataan memang negative ("tidak").
- FP = FALSE POSITIVE, model sukses memprediksi positif ("ya"), namun salah karena kenyataannya negatif ("tidak").
- FN = FALSE NEGATIVE, model sukses memprediksi negative ("ya"), karena kenyataan memang negative ("ya").

Maka Confusion Matrix memberikan informasi bahwa, dari nilai TP dan TN artinya model dapat memprediksi dengan benar keadaan yang seharusnya 'ya' dan 'tidak' sebanyak 10.725 dan 439 kasus. Sedangkan dari nilai FP dan FN artinya model tidak dapat memprediksi dengan benar (gagal prediksi) keadaan yang seharusnya 'ya' dan 'tidak' sebanyak 243 dan 950 kasus.

Kita juga mendapatkan insight bahwa data berlabel 0 atau 'tidak' jumlahnya lebih banyak daripada data berlabel 1 atau 'ya'.



Kinerja model bisa disebut ideal ketika FN dan FP bernilai 0 terutama misalnya untuk dataset yang rentan terhadap FN dan FP, anggap saja data perbankan atau saham. Tetapi dari sumber literatur yang saya baca keadaan ideal ini jarang terwujud apalagi ketika berhadapan dengan dataset di dunia industri.


Pada perhitungan manual sebenarnya Confusion Matrix bisa digunakan untuk perhitungan akurasi model.  Contohnya seperti ini:

- `accuracy = ((TP+TN) / (TP+TN+FP+FN))`

Kita coba buktikan nilai akurasi dengan cara perhitungan manual.

    accuracy = ((10725+439)/(10725+439+243+950))
             = 11.164/12.357
             = 0,90345553127781824067330258153273
             = 0,90

Jika kita meilhat classification report bisa kita lihat angka accuracy-nya sama seperti hasil perhitungan manual akurasi menggunakan Confusion Matrix yaitu sebesar 0,90.



**1.2.Classification Report:**
              
              precision    recall  f1-score   support

           0       0.92      0.98      0.95     10968
           1       0.64      0.32      0.42      1389

    accuracy                            0.90     12357
    macro avg       0.78      0.65      0.69     12357
    weighted avg    0.89      0.90      0.89     12357


a. Precision dan Recall dalam Konteks Bisnis atau Domain

- Kelas 0 (Mayoritas): Karena kelas 0 memiliki precision yang tinggi (0.92), ini menunjukkan bahwa model sangat berhati-hati dalam memastikan data kelas 0 tidak salah diprediksi sebagai kelas 1. Di beberapa aplikasi, seperti pada kasus dengan risiko finansial rendah atau moderat, ini mungkin cukup memadai.

- Kelas 1 (Minoritas): Recall rendah (0.32) pada kelas 1 menunjukkan model tidak dapat menangkap sebagian besar data yang sebenarnya termasuk dalam kelas 1. Jika kelas 1 berhubungan dengan kejadian berisiko tinggi (seperti penipuan atau deteksi penyakit serius), recall rendah ini berarti banyak kasus kritis yang terlewat, berpotensi merugikan. Precision yang rendah di kelas 1 juga menunjukkan banyak data yang sebenarnya kelas 0 justru dianggap sebagai kelas 1, yang bisa berarti peningkatan jumlah false alarms.

b. Implikasi dari Macro Average dan Weighted Average

- Macro Average (Precision: 0.78, Recall: 0.65, F1-Score: 0.69): Rata-rata ini menghitung rata-rata dari semua kelas tanpa memperhitungkan jumlah data di setiap kelas, memberi pandangan lebih netral tentang performa keseluruhan. Nilai precision dan recall yang moderat (0.78 dan 0.65) di macro average ini mengindikasikan bahwa model belum optimal dalam menangani kelas minoritas dan menunjukkan kinerja yang bias ke kelas mayoritas.

- Weighted Average (Precision: 0.89, Recall: 0.90, F1-Score: 0.89): Rata-rata tertimbang ini, di sisi lain, menunjukkan hasil yang lebih tinggi karena didominasi oleh performa di kelas 0. Karena kelas 0 jauh lebih banyak jumlahnya, weighted average ini mungkin tampak memberikan performa yang baik, tetapi perlu diingat bahwa kelas 1 tetap bermasalah.


c. Analisis Kesalahan (Error Analysis)

- False Positives (FP) dan False Negatives (FN): Dari matriks konfusi, kita tahu terdapat 243 kesalahan False Positives untuk kelas 0 dan 950 kesalahan False Negatives untuk kelas 1. Ini menunjukkan bahwa model lebih sering gagal dalam mengidentifikasi kelas 1 (yang mungkin krusial) dibandingkan kesalahan pada kelas 0. FP dan FN yang tinggi di kelas 1 bisa berimplikasi signifikan di aplikasi real-world. Misalnya, jika model diterapkan di sektor keuangan untuk mendeteksi anomali, maka 950 FN artinya ada 950 kejadian penting yang tidak terdeteksi, sementara 243 FP bisa menyebabkan alarm palsu yang menguras sumber daya untuk diperiksa.


d. Evaluasi Terhadap Akurasi
- Meskipun akurasi model mencapai 90%, hal ini dapat menyesatkan karena mayoritas prediksi yang benar adalah untuk kelas 0. Pada masalah ketidakseimbangan kelas, akurasi menjadi kurang informatif sebagai metrik kinerja. Jika tujuan utama analisis adalah untuk mendeteksi kelas minoritas (kelas 1), maka akurasi tinggi tidak serta merta menjamin model yang baik.
Untuk itu, metrik yang lebih fokus pada performa di kelas 1 (seperti F1-score kelas 1, atau recall dan precision per kelas) akan lebih relevan untuk dijadikan dasar keputusan.


#2. Latih Model SVM dengan Kernel Linear, Polynmial, RBF dan Sigmoid (Undersample)


###a. Kernel Linear Undersampling


Undersampling dataset untuk mempersingkat proses training SVM.


In [9]:
!pip install imbalanced-learn #Instalasi library `imbalanced-learn`

In [10]:
#Mendefinisikan ulang x dan y setelah undersampling.
x = data.drop('y', axis=1)
y = data['y']

In [11]:
from imblearn.under_sampling import RandomUnderSampler

# c. Buat objek RandomUnderSampler
undersampler = RandomUnderSampler(random_state=42, sampling_strategy= "auto")

# d. Lakukan undersampling pada data
x_resampled, y_resampled = undersampler.fit_resample(x, y)

In [12]:
# e. Ubah target menjadi 0 dan 1 (yes/no)
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y_resampled)

# f. Ubah fitur kategorikal menjadi numerik
X = pd.get_dummies(x_resampled, drop_first=True)

In [13]:
# Bagi data menjadi data latih dan uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalisasi fitur
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [14]:
# Buat dan latih model SVM
svm_model = SVC(kernel='linear')
svm_model.fit(X_train, y_train)

SVC(kernel='linear')

In [15]:
# Prediksi data uji
y_pred = svm_model.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1181  216]
 [ 119 1268]]

Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.85      0.88      1397
           1       0.85      0.91      0.88      1387

    accuracy                           0.88      2784
   macro avg       0.88      0.88      0.88      2784
weighted avg       0.88      0.88      0.88      2784



**Interpretasi**

**Confusion Matrix**

Kita mengukur kinerja model SVM yang menghasilkan *binary classification* menggunakan *confusion matrix*. Matriks ini memetakan kinerja model saat memproses data prediksi dibandingkan kenyataannya.

```
[[1181  216]
 [ 119 1268]]
```

Matriks ini menunjukkan hasil berikut:

```
[[TN    FP]
 [FN     TP]]
```

**Keterangan Matriks:**
- **TP = TRUE POSITIVE**: Model sukses memprediksi positif ("ya"), dan kenyataan juga positif ("ya"), sebanyak 1268 kali.
- **TN = TRUE NEGATIVE**: Model sukses memprediksi negatif ("tidak"), dan kenyataan juga negatif ("tidak"), sebanyak 1181 kali.
- **FP = FALSE POSITIVE**: Model memprediksi positif ("ya") secara salah untuk kasus yang sebenarnya negatif ("tidak") sebanyak 216 kali.
- **FN = FALSE NEGATIVE**: Model memprediksi negatif ("tidak") secara salah untuk kasus yang sebenarnya positif ("ya") sebanyak 119 kali.

Dari matriks ini, terlihat bahwa model cukup baik dalam memprediksi kedua kelas, meskipun masih terjadi kesalahan pada FP (216) dan FN (119).

**Classification Report:**
```
              precision    recall  f1-score   support

           0       0.91      0.85      0.88      1397
           1       0.85      0.91      0.88      1387

    accuracy                           0.88      2784
   macro avg       0.88      0.88      0.88      2784
weighted avg       0.88      0.88      0.88      2784
```

a. Precision dan Recall dalam Konteks Bisnis atau Domain
- **Kelas 0 (Negatif)**: Dengan **precision sebesar 0.91**, model sangat berhati-hati dalam memprediksi kelas 0, yang berarti sebagian besar prediksi "tidak" benar. **Recall 0.85** menunjukkan bahwa model bisa menangkap sebagian besar data yang seharusnya negatif dengan baik.
  
- **Kelas 1 (Positif)**: **Recall sebesar 0.91** menunjukkan bahwa model dapat menangkap sebagian besar data yang benar-benar kelas 1, yang penting jika kelas 1 terkait dengan kasus-kasus krusial. Precision di kelas 1 sebesar 0.85 juga cukup tinggi, namun masih ada kesalahan yang menghasilkan *false alarms* atau *false positives*.

b. Implikasi dari Macro Average dan Weighted Average
- **Macro Average (Precision: 0.88, Recall: 0.88, F1-Score: 0.88)**: Nilai rata-rata ini menghitung performa setiap kelas tanpa memperhitungkan ukuran kelas, memberikan gambaran yang seimbang antara kedua kelas.
  
- **Weighted Average (Precision: 0.88, Recall: 0.88, F1-Score: 0.88)**: Rata-rata tertimbang menunjukkan nilai yang sama, karena kedua kelas seimbang dalam ukuran data.

c. Analisis Kesalahan (Error Analysis)
- **False Positives (FP) dan False Negatives (FN)**: Matriks konfusi menunjukkan 216 *false positives* di kelas 0 dan 119 *false negatives* di kelas 1. Meskipun kesalahan ini relatif kecil, penting untuk diperhatikan pada kasus-kasus sensitif. Kesalahan FP bisa menambah beban karena alarm palsu, dan kesalahan FN bisa berarti kejadian penting yang tidak terdeteksi.

d. Evaluasi Terhadap Akurasi
- **Akurasi Keseluruhan (0.88)**: Akurasi 88% menunjukkan performa yang cukup baik, namun penting untuk juga memperhatikan precision, recall, dan f1-score untuk masing-masing kelas. Pada konteks ketidakseimbangan kelas, akurasi yang tinggi mungkin kurang representatif, namun pada kasus ini, kelas seimbang sehingga akurasi dapat menjadi metrik yang valid.

Model ini memiliki performa yang seimbang dan dapat digunakan secara efektif dengan hanya sedikit perbaikan dalam mengurangi kesalahan di FP dan FN.

###b. Kernel Polynomial Undersampling

In [16]:
# Buat dan latih model SVM
svm_model = SVC(kernel='poly')
svm_model.fit(X_train, y_train)

SVC(kernel='poly')

In [17]:
# Prediksi data uji
y_pred = svm_model.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1199  198]
 [ 202 1185]]

Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.86      0.86      1397
           1       0.86      0.85      0.86      1387

    accuracy                           0.86      2784
   macro avg       0.86      0.86      0.86      2784
weighted avg       0.86      0.86      0.86      2784



**Interpretasi**

**Confusion Matrix**

Matriks konfusi digunakan untuk mengevaluasi kinerja model SVM yang melakukan klasifikasi biner. Matriks ini menunjukkan bagaimana prediksi model dibandingkan dengan nilai sebenarnya.

```
[[1199  198]
 [ 202 1185]]
```

Di mana:

```
[[TN    FP]
 [FN     TP]]
```

**Keterangan Matriks:**
- **TP (True Positive)** = 1185: Model sukses memprediksi positif untuk kelas 1 ketika kenyataannya memang kelas 1.
- **TN (True Negative)** = 1199: Model sukses memprediksi negatif untuk kelas 0 ketika kenyataannya memang kelas 0.
- **FP (False Positive)** = 198: Model memprediksi positif untuk kelas 1 secara salah ketika kenyataannya adalah kelas 0.
- **FN (False Negative)** = 202: Model memprediksi negatif untuk kelas 0 secara salah ketika kenyataannya adalah kelas 1.

Dari sini, kita melihat bahwa model lebih akurat dalam memprediksi kelas yang sebenarnya (TP dan TN) dibandingkan dengan kesalahan prediksi (FP dan FN).

**Classification Report:**
```
              precision    recall  f1-score   support

           0       0.86      0.86      0.86      1397
           1       0.86      0.85      0.86      1387

    accuracy                           0.86      2784
   macro avg       0.86      0.86      0.86      2784
weighted avg       0.86      0.86      0.86      2784
```

#### a. Precision dan Recall dalam Konteks Bisnis atau Domain
- **Kelas 0 (Negatif)**: Precision dan recall untuk kelas 0 adalah 0.86, menunjukkan bahwa model memiliki kinerja yang baik dalam memastikan prediksi negatif. Model cukup andal dalam mengenali data kelas 0 dengan tepat.
  
- **Kelas 1 (Positif)**: Precision dan recall untuk kelas 1 masing-masing juga sekitar 0.86 dan 0.85. Meskipun sedikit lebih rendah dibandingkan kelas 0, model masih bisa menangkap sebagian besar data yang benar-benar kelas 1 secara efektif.

#### b. Implikasi dari Macro Average dan Weighted Average
- **Macro Average (Precision: 0.86, Recall: 0.86, F1-Score: 0.86)**: Macro average memberikan gambaran yang seimbang antar kelas karena keduanya memiliki ukuran data yang hampir sama.
  
- **Weighted Average (Precision: 0.86, Recall: 0.86, F1-Score: 0.86)**: Weighted average menghasilkan nilai yang sama dengan macro average karena kelas-kelasnya relatif seimbang.

#### c. Analisis Kesalahan (Error Analysis)
- **False Positives (FP) dan False Negatives (FN)**: Terdapat 198 false positives untuk kelas 0 dan 202 false negatives untuk kelas 1. Meski nilainya tidak terlalu besar, kesalahan ini bisa berarti bahwa model salah mengklasifikasikan data kelas positif sebagai negatif atau sebaliknya. Jika model ini diterapkan pada domain yang memerlukan presisi tinggi, mengurangi kesalahan ini akan signifikan dalam meningkatkan akurasi keseluruhan.

#### d. Evaluasi Terhadap Akurasi
- **Akurasi Keseluruhan (0.86)**: Akurasi 86% menunjukkan bahwa model cukup andal secara keseluruhan, dengan performa seimbang di antara kedua kelas. Karena kedua kelas memiliki ukuran data yang mirip, akurasi merupakan metrik yang valid untuk mengevaluasi model ini.

Model ini memiliki performa yang baik dan seimbang, namun perbaikan tambahan dapat dipertimbangkan untuk mengurangi kesalahan pada FP dan FN, terutama jika kelas minoritas memiliki dampak yang signifikan dalam penerapan di dunia nyata.

###c. Kernel RBF Undersampling


In [18]:
# Buat dan latih model SVM
svm_model = SVC(kernel='rbf')
svm_model.fit(X_train, y_train)

SVC()

In [19]:
# Prediksi data uji
y_pred = svm_model.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1179  218]
 [ 129 1258]]

Classification Report:
              precision    recall  f1-score   support

           0       0.90      0.84      0.87      1397
           1       0.85      0.91      0.88      1387

    accuracy                           0.88      2784
   macro avg       0.88      0.88      0.88      2784
weighted avg       0.88      0.88      0.88      2784



**Interpretasi**

**Confusion Matrix**

Confusion Matrix memberikan gambaran mengenai performa model dalam melakukan *binary classification*.

```
[[1179  218]
 [ 129 1258]]
```

Dengan interpretasi sebagai berikut:

```
[[TN    FP]
 [FN     TP]]
```

**Keterangan Matriks:**
- **TP (True Positive)** = 1258: Model berhasil memprediksi kelas 1 (positif) dengan benar sebanyak 1258 kali.
- **TN (True Negative)** = 1179: Model berhasil memprediksi kelas 0 (negatif) dengan benar sebanyak 1179 kali.
- **FP (False Positive)** = 218: Model secara salah memprediksi kelas 1 ketika kenyataannya kelas 0.
- **FN (False Negative)** = 129: Model secara salah memprediksi kelas 0 ketika kenyataannya kelas 1.

Dari matriks ini, dapat dilihat bahwa model lebih sering benar dalam memprediksi kelas yang benar daripada salah, namun tetap terdapat kesalahan pada FP dan FN.


**Classification Report:**
```
              precision    recall  f1-score   support

           0       0.90      0.84      0.87      1397
           1       0.85      0.91      0.88      1387

    accuracy                           0.88      2784
   macro avg       0.88      0.88      0.88      2784
weighted avg       0.88      0.88      0.88      2784
```

#### a. Precision dan Recall dalam Konteks Bisnis atau Domain
- **Kelas 0 (Negatif)**: Precision sebesar 0.90 menunjukkan bahwa ketika model memprediksi kelas 0, ia benar dalam 90% dari kasus. Recall sebesar 0.84 menunjukkan bahwa model menangkap 84% dari semua kasus yang seharusnya kelas 0.
  
- **Kelas 1 (Positif)**: Dengan precision sebesar 0.85, model akurat dalam 85% prediksi untuk kelas 1. Recall sebesar 0.91 menunjukkan bahwa model menangkap 91% dari semua kasus yang benar-benar kelas 1.

#### b. Implikasi dari Macro Average dan Weighted Average
- **Macro Average (Precision: 0.88, Recall: 0.88, F1-Score: 0.88)**: Memberikan gambaran performa keseluruhan antar kelas tanpa memperhitungkan ukuran kelas, dan mencerminkan performa seimbang antar kelas.
  
- **Weighted Average (Precision: 0.88, Recall: 0.88, F1-Score: 0.88)**: Weighted average mengonfirmasi performa yang baik dengan mempertimbangkan distribusi kelas, yang hampir seimbang.

#### c. Analisis Kesalahan (Error Analysis)
- **False Positives (FP) dan False Negatives (FN)**: Terdapat 218 *false positives* dan 129 *false negatives*. Kesalahan ini signifikan jika kelas positif memiliki dampak kritis, misalnya pada kasus-kasus risiko tinggi. Mengurangi jumlah kesalahan FN akan meningkatkan deteksi kejadian penting dalam kelas 1.

#### d. Evaluasi Terhadap Akurasi
- **Akurasi Keseluruhan (0.88)**: Akurasi 88% menunjukkan bahwa model dapat diandalkan dengan performa yang relatif kuat pada kedua kelas, memberikan hasil yang baik untuk data yang relatif seimbang.

Model ini bekerja dengan baik dan memberikan hasil yang seimbang, tetapi mengurangi FP dan FN, terutama dalam konteks yang sensitif, dapat lebih meningkatkan kinerjanya.

###d. Kernel Sigmoid Undersampling


In [20]:
# Buat dan latih model SVM
svm_model = SVC(kernel='sigmoid')
svm_model.fit(X_train, y_train)

SVC(kernel='sigmoid')

In [21]:
# Prediksi data uji
y_pred = svm_model.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1142  255]
 [ 208 1179]]

Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.82      0.83      1397
           1       0.82      0.85      0.84      1387

    accuracy                           0.83      2784
   macro avg       0.83      0.83      0.83      2784
weighted avg       0.83      0.83      0.83      2784



**Interpretasi**

**Confusion Matrix**

Matriks konfusi mengukur performa model SVM dalam klasifikasi biner, memperlihatkan prediksi model dibandingkan dengan kenyataannya.

```
[[1142  255]
 [ 208 1179]]
```

Interpretasinya:

```
[[TN    FP]
 [FN     TP]]
```

**Keterangan Matriks:**
- **TP (True Positive)** = 1179: Model benar memprediksi kelas 1 (positif) sebanyak 1179 kali.
- **TN (True Negative)** = 1142: Model benar memprediksi kelas 0 (negatif) sebanyak 1142 kali.
- **FP (False Positive)** = 255: Model salah memprediksi kelas 1 untuk data kelas 0 sebanyak 255 kali.
- **FN (False Negative)** = 208: Model salah memprediksi kelas 0 untuk data kelas 1 sebanyak 208 kali.

Dari matriks ini, dapat disimpulkan bahwa model lebih sering melakukan kesalahan dalam bentuk false positives dan false negatives.

**Classification Report:**
```
              precision    recall  f1-score   support

           0       0.85      0.82      0.83      1397
           1       0.82      0.85      0.84      1387

    accuracy                           0.83      2784
   macro avg       0.83      0.83      0.83      2784
weighted avg       0.83      0.83      0.83      2784
```

#### a. Precision dan Recall dalam Konteks Bisnis atau Domain
- **Kelas 0 (Negatif)**: Precision 0.85 mengindikasikan bahwa 85% dari prediksi untuk kelas 0 benar. Recall 0.82 menunjukkan bahwa model menangkap 82% dari seluruh kasus yang benar-benar kelas 0.
  
- **Kelas 1 (Positif)**: Precision sebesar 0.82 menunjukkan akurasi 82% dalam prediksi kelas 1, sementara recall sebesar 0.85 menunjukkan bahwa model menangkap 85% dari semua kasus yang sebenarnya kelas 1.

#### b. Implikasi dari Macro Average dan Weighted Average
- **Macro Average (Precision: 0.83, Recall: 0.83, F1-Score: 0.83)**: Macro average memberikan pandangan keseluruhan yang seimbang antar kelas dan menunjukkan bahwa model memiliki performa yang hampir setara di antara kelas-kelasnya.
  
- **Weighted Average (Precision: 0.83, Recall: 0.83, F1-Score: 0.83)**: Karena kedua kelas hampir setara dalam jumlah, weighted average memperkuat bahwa performa model cukup baik di kedua kelas.

#### c. Analisis Kesalahan (Error Analysis)
- **False Positives (FP) dan False Negatives (FN)**: Model menghasilkan 255 *false positives* dan 208 *false negatives*, menunjukkan bahwa model cenderung salah dalam memprediksi baik kelas 0 maupun kelas 1. Untuk aplikasi yang sensitif, penurunan nilai FN atau FP sangat penting, misalnya dalam skenario yang membutuhkan presisi tinggi.

#### d. Evaluasi Terhadap Akurasi
- **Akurasi Keseluruhan (0.83)**: Akurasi 83% menunjukkan bahwa model memiliki performa yang memadai meskipun ada potensi peningkatan, terutama dalam mengurangi FN dan FP untuk meningkatkan hasil klasifikasi pada kedua kelas.

Kesimpulannya, model ini menunjukkan kinerja yang seimbang antara kedua kelas tetapi dapat ditingkatkan lebih lanjut dengan mengurangi jumlah kesalahan FN dan FP, terutama dalam aplikasi di mana kesalahan memiliki konsekuensi besar.

#2.1. Berdasarkan f1-score dari hasil evaluasi, tentukan kernel mana yang memberikan hasil terbaik?

Kernel dengan hasil terbaik adalah Linear, RBF, Poly dan Sigmoid.

#3. Menyesuaikan degree terbaik untuk Kernel Polynomial

Tentukan nilai degree terbaik untuk hasil terbaik.

Degree = 2

In [22]:
svm_model_d2= SVC(kernel='poly', degree=2)
svm_model_d2.fit(X_train, y_train)

SVC(degree=2, kernel='poly')

In [23]:
# Prediksi data uji
y_pred = svm_model_d2.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1213  184]
 [ 318 1069]]

Classification Report:
              precision    recall  f1-score   support

           0       0.79      0.87      0.83      1397
           1       0.85      0.77      0.81      1387

    accuracy                           0.82      2784
   macro avg       0.82      0.82      0.82      2784
weighted avg       0.82      0.82      0.82      2784



Degree = 3

In [24]:
svm_model_d3= SVC(kernel='poly', degree=3)
svm_model_d3.fit(X_train, y_train)

# Prediksi data uji
y_pred = svm_model_d3.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1199  198]
 [ 202 1185]]

Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.86      0.86      1397
           1       0.86      0.85      0.86      1387

    accuracy                           0.86      2784
   macro avg       0.86      0.86      0.86      2784
weighted avg       0.86      0.86      0.86      2784



Degree = 4

In [25]:
svm_model_d4= SVC(kernel='poly', degree=4)
svm_model_d4.fit(X_train, y_train)

# Prediksi data uji
y_pred = svm_model_d4.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1247  150]
 [ 449  938]]

Classification Report:
              precision    recall  f1-score   support

           0       0.74      0.89      0.81      1397
           1       0.86      0.68      0.76      1387

    accuracy                           0.78      2784
   macro avg       0.80      0.78      0.78      2784
weighted avg       0.80      0.78      0.78      2784



Degree = 5

In [26]:
svm_model_d5= SVC(kernel='poly', degree=5)
svm_model_d5.fit(X_train, y_train)

# Prediksi data uji
y_pred = svm_model_d5.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1281  116]
 [ 597  790]]

Classification Report:
              precision    recall  f1-score   support

           0       0.68      0.92      0.78      1397
           1       0.87      0.57      0.69      1387

    accuracy                           0.74      2784
   macro avg       0.78      0.74      0.74      2784
weighted avg       0.78      0.74      0.74      2784



Kesimpulan, kombinasi degree terbaik untuk menghasilkan classification terbaik adalah ketika kita menggunakan degree 3 pada kernel poly model ini. Penggunaan degree 3 menghasilkan nilai accuracy dan F1 sebesar 86%.

#4. Mencoba nilai C (misalnya 0.1, 1, dan 10) untuk Kernel RBF
Tentukan nilai C terbaik berdasarkan f1-score.

C = 0.1

In [28]:
svm_model_c1 = SVC(kernel='rbf', C = 0.1)
svm_model_c1.fit(X_train, y_train)

# Prediksi data uji
y_pred = svm_model_c1.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1177  220]
 [ 172 1215]]

Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.84      0.86      1397
           1       0.85      0.88      0.86      1387

    accuracy                           0.86      2784
   macro avg       0.86      0.86      0.86      2784
weighted avg       0.86      0.86      0.86      2784



C = 1

In [29]:
svm_model_c2 = SVC(kernel='rbf', C = 1)
svm_model_c2.fit(X_train, y_train)

# Prediksi data uji
y_pred = svm_model_c2.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1179  218]
 [ 129 1258]]

Classification Report:
              precision    recall  f1-score   support

           0       0.90      0.84      0.87      1397
           1       0.85      0.91      0.88      1387

    accuracy                           0.88      2784
   macro avg       0.88      0.88      0.88      2784
weighted avg       0.88      0.88      0.88      2784



C = 10

In [30]:
svm_model_c3 = SVC(kernel='rbf', C = 10)
svm_model_c3.fit(X_train, y_train)

# Prediksi data uji
y_pred = svm_model_c3.predict(X_test)

# Evaluasi hasil
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[1176  221]
 [ 170 1217]]

Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.84      0.86      1397
           1       0.85      0.88      0.86      1387

    accuracy                           0.86      2784
   macro avg       0.86      0.86      0.86      2784
weighted avg       0.86      0.86      0.86      2784



**Kesimpulan**

Menggunakan C = 1 dapat menghasilkan F1-Score tertinggi yaitu sebesar 88%.

#5. Interpretasi Soal Nomor (3) dan (4)

###Interpretasi nomor 3 dengan degree terbaik yaitu 3


**Confusion Matrix**

Matriks konfusi digunakan untuk mengevaluasi kinerja model SVM dalam klasifikasi biner. Matriks ini membandingkan prediksi model dengan nilai sebenarnya.

```
[[1281  116]
 [ 597  790]]
```

Di mana:

```
[[TN    FP]
 [FN     TP]]
```

Keterangan Matriks:

- **TP (True Positive)** = 790: Model berhasil memprediksi kelas 1 (positif) dengan benar sebanyak 790 kali.
- **TN (True Negative)** = 1281: Model berhasil memprediksi kelas 0 (negatif) dengan benar sebanyak 1281 kali.
- **FP (False Positive)** = 116: Model memprediksi kelas 1 (positif) secara salah untuk data yang sebenarnya adalah kelas 0 (negatif) sebanyak 116 kali.
- **FN (False Negative)** = 597: Model memprediksi kelas 0 (negatif) secara salah untuk data yang sebenarnya adalah kelas 1 (positif) sebanyak 597 kali.

Dari matriks ini, kita dapat melihat bahwa model menunjukkan kesulitan dalam mendeteksi kelas 1 (positif), dengan FN yang cukup tinggi.

**Classification Report**

```
              precision    recall  f1-score   support

           0       0.68      0.92      0.78      1397
           1       0.87      0.57      0.69      1387

    accuracy                           0.74      2784
   macro avg       0.78      0.74      0.74      2784
weighted avg       0.78      0.74      0.74      2784
```

a. Precision dan Recall dalam Konteks Bisnis atau Domain
- **Kelas 0 (Negatif)**: Precision 0.68 dan recall 0.92 menunjukkan bahwa meskipun model cukup baik dalam mengidentifikasi data negatif (dari 1397 total data negatif, 1281 benar-benar diprediksi sebagai negatif), ada risiko kesalahan dalam memprediksi kelas ini sebagai positif (FP yang tinggi).

- **Kelas 1 (Positif)**: Precision 0.87 menunjukkan bahwa model dapat mendeteksi kelas positif dengan baik, tetapi dengan recall 0.57, banyak data kelas 1 yang terlewat (597 FN). Ini berarti model tidak dapat menangkap sebagian besar data yang seharusnya teridentifikasi sebagai positif.

b. Implikasi dari Macro Average dan Weighted Average
- **Macro Average (Precision: 0.78, Recall: 0.74, F1-Score: 0.74)**: Rata-rata ini menunjukkan bahwa model memiliki kinerja yang moderat secara keseluruhan. Nilai recall yang lebih rendah di kelas 1 menunjukkan bahwa model cenderung lebih menguntungkan kelas 0.

- **Weighted Average (Precision: 0.78, Recall: 0.74, F1-Score: 0.74)**: Rata-rata tertimbang ini menunjukkan hasil yang serupa dengan macro average, tetapi dipengaruhi oleh proporsi masing-masing kelas dalam dataset. Hal ini menekankan bahwa meskipun ada performa yang baik di kelas 0, performa pada kelas 1 perlu ditingkatkan.

c. Analisis Kesalahan (Error Analysis)
- **False Positives (FP) dan False Negatives (FN)**: Kesalahan model tercermin dari 116 FP dan 597 FN. Tingginya FN menunjukkan bahwa model sering kali gagal mendeteksi kasus positif, yang mungkin berarti model kurang sensitif terhadap kelas 1. Ini bisa menjadi masalah di domain di mana kelas 1 sangat penting, seperti dalam deteksi penyakit atau deteksi penipuan.

d. Evaluasi Terhadap Akurasi
- **Akurasi Keseluruhan (0.74)**: Akurasi 74% menunjukkan bahwa model tidak sepenuhnya efektif, terutama karena ketidakseimbangan antara kelas. Dalam konteks ketidakseimbangan kelas, akurasi mungkin tidak mencerminkan kinerja model secara akurat. Dengan FN yang tinggi pada kelas 1, model perlu disesuaikan untuk meningkatkan kemampuannya dalam mendeteksi kelas ini.

Model SVM ini menunjukkan potensi, tetapi ada ruang untuk perbaikan, khususnya dalam mendeteksi kelas positif. Penggunaan kernel polynomial dengan derajat 3 berhasil menghasilkan akurasi dan F1 score yang lebih baik, tetapi perlu diteliti lebih lanjut untuk mengatasi isu FN yang tinggi pada kelas 1. Dalam pengembangan lebih lanjut, mempertimbangkan teknik pengimbangan data atau tuning parameter lebih lanjut dapat membantu meningkatkan performa model secara keseluruhan.

Maka dengan ini penggunaan degree 3 menghasilkan F1 model terbaik yaitu sebesar 86% dibandingkan degree lainnya. Sementara degree 2 menghasilkan 82%, degree 4 menghasilkan 78% dan degree 5 menghasilkan 74%.

###Interpretasi nomor 4 dengan C terbaik yaitu 2

**Confusion Matrix**

Matriks konfusi ini digunakan untuk mengevaluasi kinerja model SVM dengan kernel Radial Basis Function (RBF) dan parameter C = 1. Berikut adalah matriks yang dihasilkan:

```
[[1179  218]
 [ 129 1258]]
```

Matriks ini dapat diinterpretasikan sebagai berikut:

```
[[TN    FP]
 [FN     TP]]
```

**Keterangan Matriks:**
- **TP (True Positive)** = 1258: Model berhasil memprediksi kelas 1 (positif) dengan benar sebanyak 1258 kali.
- **TN (True Negative)** = 1179: Model berhasil memprediksi kelas 0 (negatif) dengan benar sebanyak 1179 kali.
- **FP (False Positive)** = 218: Model memprediksi kelas 1 (positif) secara salah untuk data yang sebenarnya kelas 0 (negatif) sebanyak 218 kali.
- **FN (False Negative)** = 129: Model memprediksi kelas 0 (negatif) secara salah untuk data yang sebenarnya kelas 1 (positif) sebanyak 129 kali.

Dari matriks ini, kita dapat melihat bahwa model relatif baik dalam memprediksi kelas positif (kelas 1), tetapi masih ada beberapa kesalahan dalam mendeteksi kelas negatif (kelas 0).

**1.2. Classification Report**

**Classification Report:**
```
              precision    recall  f1-score   support

           0       0.90      0.84      0.87      1397
           1       0.85      0.91      0.88      1387

    accuracy                           0.88      2784
   macro avg       0.88      0.88      0.88      2784
weighted avg       0.88      0.88      0.88      2784
```

a. Precision dan Recall dalam Konteks Bisnis atau Domain
- **Kelas 0 (Negatif)**: Precision 0.90 menunjukkan bahwa model memiliki tingkat akurasi yang baik dalam memprediksi kelas negatif (dari 1397 total data negatif, 1179 diprediksi dengan benar). Namun, dengan recall 0.84, ada 218 kasus yang salah diprediksi sebagai positif (False Positives), yang menunjukkan bahwa model mungkin terlalu konservatif dan menyebabkan kesalahan di kelas ini.

- **Kelas 1 (Positif)**: Precision 0.85 menunjukkan bahwa model memiliki kemampuan yang baik untuk memprediksi positif, meskipun terdapat 129 kasus yang salah diprediksi sebagai negatif (False Negatives). Recall 0.91 menunjukkan bahwa model cukup baik dalam mendeteksi kelas 1, tetapi masih ada ruang untuk perbaikan.

b. Implikasi dari Macro Average dan Weighted Average
- **Macro Average (Precision: 0.88, Recall: 0.88, F1-Score: 0.88)**: Rata-rata ini menghitung rata-rata dari semua kelas tanpa memperhitungkan jumlah data di setiap kelas. Nilai ini menunjukkan bahwa model secara umum berfungsi dengan baik dan memiliki keseimbangan yang baik antara precision dan recall untuk kedua kelas.

- **Weighted Average (Precision: 0.88, Recall: 0.88, F1-Score: 0.88)**: Rata-rata tertimbang ini memperhitungkan jumlah contoh dalam setiap kelas, menunjukkan bahwa performa model cenderung stabil di kedua kelas, meskipun kelas positif sedikit lebih sulit untuk diprediksi.

c. Analisis Kesalahan (Error Analysis)
- **False Positives (FP) dan False Negatives (FN)**: Model ini menghasilkan 218 FP dan 129 FN. Tingginya jumlah FP menunjukkan bahwa model cenderung salah mengidentifikasi data negatif sebagai positif, yang dapat menyebabkan pemborosan sumber daya dalam situasi di mana prediksi positif perlu divalidasi lebih lanjut. Di sisi lain, 129 FN menunjukkan bahwa model melewatkan beberapa kasus positif, yang bisa menjadi masalah di konteks di mana kelas 1 memiliki risiko tinggi, seperti dalam deteksi penyakit atau penipuan.

d. Evaluasi Terhadap Akurasi
- **Akurasi Keseluruhan (0.88)**: Akurasi 88% menunjukkan bahwa model cukup baik dalam memprediksi baik kelas 0 maupun kelas 1. Namun, seperti sebelumnya, akurasi tidak cukup menggambarkan kinerja model dalam konteks ketidakseimbangan kelas. Penting untuk menilai metrik lain seperti precision dan recall untuk masing-masing kelas, yang memberikan gambaran lebih jelas tentang kinerja model.

**Kesimpulan**

Model SVM dengan kernel RBF dan parameter C = 1 menunjukkan kinerja yang terbaik dengan akurasi 88% dibandingkan nilai C lainnya. Meskipun model ini cukup baik dalam mendeteksi kelas 1, masih ada ruang untuk perbaikan, terutama dalam mengurangi jumlah False Positives dan False Negatives. Dalam pengembangan lebih lanjut, strategi seperti pengimbangan data atau tuning hyperparameter lebih lanjut dapat membantu meningkatkan deteksi kelas positif, yang sangat penting dalam banyak aplikasi bisnis.

Sementara nilai C=0.1 dan C = 10 sama-sama menghasilkan F1-Score sebesar 86%.

#6. Mengapa Kernel Tertentu Mungkin Lebih Cocok

- Kernel Polynomial berfungsi baik dalam beberapa situasi, terutama ketika hubungan antara fitur dan kelas tidak linear, tetapi pemilihan derajat yang tepat sangat penting. Pada derajat yang lebih tinggi, model bisa menjadi terlalu kompleks dan menyebabkan overfitting, seperti yang terlihat pada degree 4 dan 5.

- Kernel RBF cenderung lebih fleksibel dan sering kali lebih baik dalam menangkap pola kompleks dalam data. Dengan pengaturan parameter C yang tepat, model dapat mengontrol margin dan kesalahan klasifikasi, sehingga dapat menghasilkan performa yang lebih baik dalam banyak kasus.

- Dalam dataset ini, kernel RBF dengan C = 1 memberikan hasil yang lebih konsisten dan lebih tinggi pada F1-Score, menunjukkan bahwa ia lebih mampu menangkap kompleksitas dalam data dibandingkan dengan kernel Polynomial pada derajat yang lebih tinggi.